lbrynet/core/server/DHTHashAnnouncer.py

```diff
@@ -87,12 +90,32 @@ def announce():
 
 
 class DHTHashSupplier(object):
+    # 1 hour is the min time hash will be reannounced
+    MIN_HASH_REANNOUNCE_TIME = 60*60
+    # conservative assumption of the time it takes to announce
+    # a single hash
+    SINGLE_HASH_ANNOUNCE_DURATION = 1
+
     """Classes derived from this class give hashes to a hash announcer"""
     def __init__(self, announcer):
         if announcer is not None:
             announcer.add_supplier(self)
         self.hash_announcer = announcer
-        self.hash_reannounce_time = 60 * 60  # 1 hour
 
     def hashes_to_announce(self):
         pass
+
+    """
+    Hash reannounce time is set to current time + MIN_HASH_REANNOUNCE_TIME,
+    unless we are announcing a lot of hashes at once which could cause the
+    the announce queue to pile up.  To prevent pile up, reannounce
+    only after a conservative estimate of when it will finish
+    to announce all the hashes.
+    """
+    def get_next_announce_time(self, num_hashes_to_announce=1):
+        queue_size = self.hash_announcer.hash_queue_size()+num_hashes_to_announce
+        reannounce = max(self.MIN_HASH_REANNOUNCE_TIME,
+                            queue_size*self.SINGLE_HASH_ANNOUNCE_DURATION)
+        return time.time() + reannounce
```

Comment:
I'm not convinced that this is a totally effective solution as it largely moves the queue from being in the queue data structure to being in the database.  The main difference is that the database will only have on entry for each blob so you decrease the chances of having repeated items in the queue - which is an improvement.

A better long term solution would be something along the lines of https://github.com/lbryio/lbry/pull/435/files - which is to record when a blob was last announced (instead of when to next announce it) and have a data structure that doesn't allow for repeated blobs.  As that requires a more substantial change, I'm okay with putting this in place for now but we should follow up with a more complete fix.